### PR TITLE
Feature/API V2 Tokens and Helpers

### PR DESCRIPTION
**Pull request by brianjgeiger**
Purpose
-----------
Make it easier for developers to use the V2 API by providing token-based authentication (in preparation for full OAUTH2 when the UI is merged in) and some helper functions to generate proper URLs in development that won't break on staging/production.

Changes
-------------
Add a new authentication class for CAS based token authentication. Add helper functions in JS and Python to get the base URL of the API from the settings and use that instead of hardcoding the domain, the port, and the base path. And, of course, tests.

Side effects
----------------
No side effects anticipated, though a new authentication scheme always has potential for excitement.

### CODE REVIEW

File: tests/api_tests/base/test_auth.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+        self.user2 = UserFactory()
+
+        # Test projects for which a given user DOES and DOES NOT  have appropriate permissions
+        self.reachable_project = ProjectFactory(title="Private Project User 1", is_public=False, creator=self.user1)
+        self.unreachable_project = ProjectFactory(title="Private Project User 2", is_public=False, creator=self.user2)
+
+        self.reachable_url = "/{}nodes/{}/".format(API_BASE, self.reachable_project._id)
+        self.unreachable_url = "/{}nodes/{}/".format(API_BASE, self.unreachable_project._id)
+
+    def _make_request(self, url, access_token=None):
```
Comment (sloria) [inline tests/api_tests/base/test_auth.py]: Minor, but this might not be necessary. You can do:

```python
self.app.get(url, auth=access_token, auth_type='jwt')
```

File: tests/api_tests/base/test_auth.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+from nose.tools import *  # flake8: noqa
+
+from framework.auth import cas
+from tests.base import ApiTestCase
+from tests.factories import ProjectFactory, UserFactory
+
+from api.base.settings import API_BASE
+
+
+class TestOAuthValidation(ApiTestCase):
```
Comment (sloria) [inline tests/api_tests/base/test_auth.py]: :+1:

File: tests/test_utils.py [hunk 3]
```diff
@@ -59,6 +59,41 @@ def test_api_url_for(self):
         with self.app.test_request_context():
             assert api_url_for('dummy_view', pid='123') == '/api/v1/123/'
 
+    def test_api_v2_url_with_port(self):
+        """Handles URLs with port correctly"""
```
Comment (sloria) [inline tests/test_utils.py]: As per COSDev, docstrings for test methods are discouraged in favor of long test method names and code comments.

> Use long, descriptive names. This often obviates the need for doctrings in test methods. This also makes it easier to locate tests that fail.

https://cosdev.readthedocs.org/en/latest/process/testing.html#general-testing-guidelines

File: website/static/js/osfHelpers.js [hunk 4]
```diff
@@ -23,6 +23,35 @@ var growl = function(title, message, type) {
...
+ * Can optionally pass in an object with params (name:value) to be appended to URL. Calling as:
+ *   apiV2Url("users/4urxt/applications", {"a":1, "filter[fullname]":"lawrence"}, "https://staging2.osf.io/api/v2/")
+ * would yield the result:
+ *  "https://staging2.osf.io/api/v2/users/4urxt/applications?a=1&filter%5Bfullname%5D=lawrence"
+ * @param {String} pathString The string to be appended to the absolute base path, eg "users/4urxt"
+ * @param {Object} paramsObject (optional) An object containing parameters to add to the URL. Otherwise pass 'undefined'.
+ * @param {String} apiPrefix (optional) Manually specify the prefix used for API routes (useful for testing)
+ */
+var apiV2Url = function (pathString, paramsObject, apiPrefix){
+    apiPrefix = apiPrefix || window.contextVars.apiV2Prefix;
```
Comment (sloria) [inline website/static/js/osfHelpers.js]: It might make sense to use URI.js (which we use elsewhere on the site) to do this sort of URL manipulation.

File: website/util/__init__.py [hunk 5]
```diff
@@ -7,9 +7,9 @@
 import furl
 
 from flask import request, url_for
-from django.core.urlresolvers import reverse
 
-from website import settings
+from website import settings as osf_settings
+from api.base import settings as drf_settings
```
Comment (sloria) [inline website/util/__init__.py]: Slight misnomer: `api.base.settings` is the settings module for the whole Django app, which include DRF settings and more; maybe rename to `django_settings`.

File: website/util/__init__.py [hunk 6]
```diff
@@ -7,9 +7,9 @@
 import furl
 
 from flask import request, url_for
-from django.core.urlresolvers import reverse
 
-from website import settings
+from website import settings as osf_settings
```
Comment (sloria) [inline website/util/__init__.py]: `website_settings` might be a more appropriate name.

File: website/util/__init__.py (hunk 5 above)
Comment (brianjgeiger) [inline website/util/__init__.py]: Maybe api_settings; this might not be the last django thing we work in like this.

File: api/base/urls.py [hunk 7]
```diff
@@ -11,9 +11,14 @@
...
-    url(base_pattern, include(patterns('',
-        url(r'^$', views.root),
-        url(r'^nodes/', include('api.nodes.urls', namespace='nodes')),
-        url(r'^users/', include('api.users.urls', namespace='users')),
-        url(r'^docs/', include('rest_framework_swagger.urls')),
-    )))] + static('/static/', document_root=settings.STATIC_ROOT)
\ No newline at end of file
+    url(base_pattern,
+        include(patterns('',
+                         url(r'^$', views.root),
```
Comment (sloria) [inline api/base/urls.py]: Hm, I preferred the style before this change, as it was more consistent with the rest of the codebase, but it's not a big deal.

**Comment by sloria** [comment]
Pass finished. :point_up:

File: website/util/__init__.py (hunk 5 above)
Comment (sloria) [inline website/util/__init__.py]: Yep, that's good.

File: website/static/js/osfHelpers.js (hunk 4 above)
Comment (sloria) [inline website/static/js/osfHelpers.js]: @abought I'm going to go ahead and merge this PR; can you make this change in a separate PR?

Comment (abought) [inline website/static/js/osfHelpers.js]: Will do. Can you wait 10-15 min? I'd just like to verify that the jwt changeout in unit tests isn't modifying the access token in the auth:bearer headers.  (eg, that it's reflecting the use case in the most accurate way possible)

Comment (sloria) [inline website/static/js/osfHelpers.js]: Sure. The token isn't modified: https://github.com/sloria/webtest-plus/blob/develop/webtest_plus/app.py#L29-L31

Comment (abought) [inline website/static/js/osfHelpers.js]: Ok, token is transmitted as-is. Looks ok to me. (I was under the impression that jwt would encrypt the token in some way, but guess not)
